fineract-investor/src/main/java/org/apache/fineract/investor/service/ExternalAssetOwnersWriteServiceImpl.java

```diff
@@ -0,0 +1,229 @@
+        ExternalTransferChangedData changedData = new ExternalTransferChangedData();
+        changedData.setSettlementDate(savedExternalAssetOwnerTransfer.getSettlementDate());
+        changedData.setOwnerExternalId(savedExternalAssetOwnerTransfer.getOwner().getExternalId().getValue());
+        changedData.setTransferExternalId(savedExternalAssetOwnerTransfer.getExternalId().getValue());
+        changedData.setPurchasePriceRatio(savedExternalAssetOwnerTransfer.getPurchasePriceRatio());
+        responseData.setChanges(changedData);
+        responseData.setDateformat(fromApiJsonHelper.extractStringNamed(ExternalTransferRequestParameters.DATEFORMAT, json));
+        responseData.setLocale(fromApiJsonHelper.extractStringNamed(ExternalTransferRequestParameters.LOCALE, json));
+        return responseData;
+    }
+
+    private void validateSale(ExternalAssetOwnerTransfer externalAssetOwnerTransfer) {
+        validateSettlementDate(externalAssetOwnerTransfer);
+        validateTransferStatusForSale(externalAssetOwnerTransfer);
+        validateLoanStatus(externalAssetOwnerTransfer);
+    }
+
+    private void validateBuyBack(ExternalAssetOwnerTransfer externalAssetOwnerTransfer) {
+        validateSettlementDate(externalAssetOwnerTransfer);
+        validateTransferStatusForBuyBack(externalAssetOwnerTransfer);
+    }
+
+    private void validateSettlementDate(ExternalAssetOwnerTransfer externalAssetOwnerTransfer) {
+        if (externalAssetOwnerTransfer.getSettlementDate().isBefore(ThreadLocalContextUtil.getBusinessDate())) {
+            throw new ExternalAssetOwnerInitiateTransferException("Settlement date cannot be in the past");
+        }
+    }
+
+    private void validateLoanStatus(ExternalAssetOwnerTransfer externalAssetOwnerTransfer) {
+        if (!Objects.isNull(externalAssetOwnerTransfer.getLoanId())
+                && !loanReadPlatformService.isLoanTransferable(externalAssetOwnerTransfer.getLoanId())) {
+            throw new ExternalAssetOwnerInitiateTransferException("Loan is not in active status");
+        } else if (!Objects.isNull(externalAssetOwnerTransfer.getExternalLoanId())
+                && loanReadPlatformService.isLoanTransferable(externalAssetOwnerTransfer.getExternalLoanId())) {
+            throw new ExternalAssetOwnerInitiateTransferException("Loan is not in active status");
+        }
+    }
+
+    private void validateTransferStatusForSale(ExternalAssetOwnerTransfer externalAssetOwnerTransfer) {
+        Optional<ExternalAssetOwnerTransfer> latestTransferOptional = externalAssetOwnerTransferRepository
+                .findLatestByLoanIdOrExternalLoanId(externalAssetOwnerTransfer.getLoanId(), externalAssetOwnerTransfer.getExternalLoanId());
+        if (latestTransferOptional.isPresent()) {
+            ExternalAssetOwnerTransfer latestTransfer = latestTransferOptional.get();
+            ExternalTransferStatus latestTransferStatus = ExternalTransferStatus.valueOf(latestTransfer.getStatus());
```

Comment:
what if it is BUYBACK but the BUYBACK settlement date is later than the new SALE?

Reply:
in this case, when we recorded the BUYBACK, the loan had to be in ACTIVE state, but we cannot trigger SALE on ACTIVE transfers